### PR TITLE
Add a simple localStorage-based autosave implementation (fixes: #894)

### PR DESCRIPTION
**Pull request by wlach**

### CODE REVIEW

**Comment by wlach** [comment]
This only needs one of @hamilton or @bcolloran to review :)

File: src/tools/autosave.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { exportJsmdToString } from './jsmd-tools'
+import { setPreviousAutosave } from '../actions/actions'
+
+function getAutosaveKey(state) {
+  const documentId = (state.notebookInfo.connectionMode === 'STANDALONE') ?
```
Comment (hamilton) [inline src/tools/autosave.js]: This is throwing an error on my local version of the branch when I load a notebook from the server, for some reason. I believe you can use`connectionModeIsStandalone(state)` here, at any rate.

File: src/reducers/notebook-reducer.js [hunk 2]
```diff
@@ -95,6 +100,13 @@ const notebookReducer = (state = newNotebook(), action) => {
       )
     }
 
+    case 'REPLACE_NOTEBOOK_CONTENT': {
+      cells = action.cells.map((cell, i) =>
+        Object.assign(newCell(i, cell.cellType), cell))
+
+      return Object.assign({}, state, { cells, title: action.title })
```
[the remark this answers is not in the document]
Comment (wlach) [reply src/reducers/notebook-reducer.js]: Originally I did use newNotebook() here as a template for state, but that caused the UI to blank out. I figured we were only supposed to use that on first load?

File: src/tools/autosave.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  } else {
+    const autosaveState = getAutosaveState(autosaveKey)
+    saveAutosaveState(autosaveKey, Object.assign(autosaveState, {
+      dirtyCopy: jsmd,
+      dirtySaved: new Date().toISOString(),
+    }))
+  }
+}
+
+function subscribeToAutoSave(store) {
```
[the remark this answers is not in the document]
Comment (wlach) [reply src/tools/autosave.js]: makes sense

**Comment by wlach** [comment]
@hamilton I think this is ready for another look

**Comment by hamilton** [comment]
@wlach Great! It works.

Another thing I'm noticing, now that I'm testing it - if I run a notebook in standalone mode, then edit it, then reload the page and click "restore", and then reload again, I lose my changes.

I'm wondering if we only apply the subscribe step of overwriting what's in local storage if there has been an edit in some way. Otherwise, I end up losing my work if I don't immediately apply, or apply the change and then don't do anything other than close.

Otherwise I think this is looking good. Going to do another pass on the code itself now that I'm understanding how it feels under the fingers.

**Comment by wlach** [comment]
@hamilton that makes sense... I believe that it will re-autosave if you make any modification after restoring, but we should probably just keep the "dirty" copy outright. this shouldn't be too hard.

**Comment by hamilton** [comment]
Yes, but just to clarify, I didn't make any modification after restoring - at least not as an intentional user action. I just refreshed.

**Comment by wlach** [comment]
Fixed the issue by not clearing autosave on restore.

File: src/tools/__tests__/autosave.test.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import { updateAutosave, getAutosaveState } from '../autosave'
+import { stateFromJsmd } from '../jsmd-tools'
+import { newNotebook, newCell } from '../../editor-state-prototypes'
+
+describe('updateAutosave', () => {
+  function setUp() {
```
Comment (hamilton) [inline src/tools/__tests__/autosave.test.js]: `beforeEach` is a Jest function that does this by default, like `beforeEach(()=>{something})`.

Comment (hamilton) [reply src/tools/__tests__/autosave.test.js]: https://jestjs.io/docs/en/setup-teardown

**Comment by wlach** [comment]
Fixed the nit and merged!

**Comment by hamilton** [comment]
🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

**Pull request merged**